bugfix/20334-win-width

KacperMadej opened the pull request:
Fixed #20334, chart width exceeded container width on Windows when scrollbars were present.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206237632708354

circleci-mu commented:
### File size comparison
Sizes for compiled+gzipped (bold) and compiled files.
| | master | candidate | difference |
|-------------|-------------:|-------------:|-------------:|
| highcharts.js | **95.6 kB**<br>270.0 kB | **95.6 kB**<br>270.1 kB | **30 B**<br>116 B |
| highstock.js | **128.5 kB**<br>371.9 kB | **128.5 kB**<br>372.0 kB | **31 B**<br>116 B |
| highmaps.js | **121.0 kB**<br>347.8 kB | **121.1 kB**<br>347.9 kB | **29 B**<br>116 B |
| highcharts-gantt.js | **130.6 kB**<br>376.7 kB | **130.7 kB**<br>376.8 kB | **30 B**<br>116 B |

circleci-mu commented:
## Visual test results - Differences found
Found **13** diffing sample(s). [Please review the differences.](https://vrevs.highsoft.com/pr/20581/review)

Xillez reviewed (approved):
LGTM 👍 Seems to work nicely

TorsteinHonsi commented:
Thanks for the PR! 

Before we commit this - do we know why this broke in v11? I have problems reproducing this myself in BrowserStack, but I suspect it may have something to do with the `ResizeObserver` refactoring that we did.

For example, what happens if we remove the `chart.hasLoaded` condition here?

https://github.com/highcharts/highcharts/blob/5ebada8704aee914e06d7effe03187babd444650/ts/Core/Chart/Chart.ts#L1815

Will it magically redraw to make room for the scrollbars? If not, can we bisect the issue to a specific commit?

Maybe we can get away with a cheaper fix if we know exactly what change that broke it.

KacperMadej commented:
@TorsteinHonsi 
Yes, this is related to the `ResizeObserver` refactor.

**Some context:**

I could recreate the previously "working" case only when there was an extra reflow running that was doing nothing (except maybe wasting time and resources) in almost all charts. The behavior was never perfect as in the code reported in the issue there is `chart.reflow` on the chart's `redraw` event ([more in the comment here](https://github.com/highcharts/highcharts/issues/20334#issuecomment-1924271703)). Moving to the `load` event would be an improvement but neither works now because in this [here](https://github.com/highcharts/highcharts/pull/18770/files#diff-eec054049d33d5ccd9ad7485a735dfeb5bda93fba3e975bdd69cdff552d400fbR1649) the container size is read and updated for the reference of the reflow function, so the initial reflow doesn't detect the change. Only `chart.continerBox` updates but nothing was adjusted.

`chart.getContainerBox();` is needed because when the `ResizeObserver` starts observing it runs the callback on activation (as expected by the spec - IMO explained better [here](https://stackoverflow.com/questions/60026223/does-resizeobserver-invoke-initially-on-page-load)), so I moved it earlier in the code execution - now the width could be registered early and an instance of Renderer gets the correct size on init.

A funny thing (maybe not if you start to question your sanity and decide to continue writing a comment the next day, but still), if you set up break-points just right you will or won't get the initial callback execution. When you observe the observer closely enough it behaves differently - without break-points the DOM is updated after the `ResizeObserver` starts observing.

**In conclusion:**

> what happens if we remove the `chart.hasLoaded` condition here?

When I am testing this, the first time `runReflow` is executed `chart.hasLoaded` is already `true`, so removing it will change nothing. In chart's `reflow` function setSize doesn't happen because `oldBox` is the same as the freshly obtained `containerBox`. This is the same behavior for the code with and without the fix, but with the fix the match between `oldBox` and `containerBox` truly represents the lack of change in size (and without the fix, represents only the match between the variables).

Sorry for the wall of text. Please let me know if something should be investigated or elaborated on more.

TorsteinHonsi reviewed (approved):
Thanks for the explanation, I understand you have considered all possibilities!
